coralproject/talk
# [CORL-2381] Persist seen comments in the back-end

## What does this PR do?

Persists seen comment state in the back-end so it can be a shared state between devices/logins for the same user.

## What changes to the GraphQL/Database Schema does this PR introduce?

- Adds a `seen` flag to the `Comment` type
- Adds a `markCommentAsSeen` mutation to the `Mutators` list

## How do I test this PR?

- Create the indexes listed in the deploy notes below
- Visit the stream as a logged in user and see that you can key through or refresh the stream and have comments go yellow to white
- Load up on your phone, another tab/browser and sign in as the same user, see that seen state is persisted to your user account
- Check your mongo for records in the `seenComments` collection
- set the following index, the `30` is 30 seconds (source: https://docs.mongodb.com/manual/core/index-ttl/)
    ```
    db.seenComments.createIndex({ lastSeenAt: 1 }, { expireAfterSeconds: 30 });
    ```
- wait 30 seconds
- watch your seenComments records get cleaned up after they're 30 seconds old
 
## How do we deploy this PR?

Create the following indexes in Mongo:

```
db.seenComments.createIndex({ tenantID: 1, storyID: 1, userID: 1 });
db.seenComments.createIndex({ lastSeenAt: 1 }, { expireAfterSeconds: 604800 });
```

Where `604800` is `7 days * 24 hours * 60 minutes * 60 seconds` for a TTL in seconds of how long to keep the seenComments records before deleting them gracefully within Mongo.
